1. Basic Web Development/script.ts

```diff
@@ -0,0 +1,405 @@
+            console.error('Error:', error);
+            permissionsRoutine();
+        });
+
+    createPermissionsForm(form);
+}
+
+function submitEdit(form: HTMLElement) {
+    const type = document.querySelector('input[type=radio][name=presetE]:checked').id;
+    const text = (document.getElementById('permNameEdit') as HTMLInputElement).value;
+
+    if (!text) {
+        document.getElementById('text-alert-edit').classList.remove('no-display');
+        return;
+    } else {
+        document.getElementById('text-alert-edit').classList.add('no-display');
+    }
+
+    const index = this.PERMISSIONS.findIndex((perm: Permission) => perm.id == this.NOWEDITING.id);
+
+    this.PERMISSIONS[index].type = type.replace('Edit', '');
+    this.PERMISSIONS[index].text = text;
+
+    document.getElementById('edit-form').classList.add('no-display');
+
+    const messageJSON = JSON.stringify({index: index, type: type.replace('Edit', ''), text: text});
+
+    fetch(API_URL + '/pref/update', { method: 'PUT', headers: HEADERS, body: messageJSON })
+        .then(response => response.json())
+        .then(data => {
+            console.log(data);
+        })
+        .catch((error) => {
+            console.error('Error:', error);
+            permissionsRoutine();
+        });
+
+    createPermissionsForm(form);
+}
+
+function handleLock(lock: HTMLInputElement, popUp: Element) {
+    lock.addEventListener('change', (event) => {
+        ISLOCKED = lock.checked;
+
+        popUp.classList.remove('hidden');
+        popUp.classList.remove('showing');
+
+        if (ISLOCKED) {
+            popUp.classList.add('showing');
+        } else {
+            popUp.classList.add('hidden');
+        }
+
+    });
+}
+
+function handlePopUp(popUp: Element) {
+    if (ISLOCKED) {
+        popUp.classList.add('showing');
+        popUp.classList.remove('hidden');
+    }
+
+    popUp.addEventListener('click', () => {
+        if (!this.ISLOCKED) {
+            popUp.classList.add('showing');
+            popUp.classList.remove('hidden');
+        }
+    });
+
+    popUp.addEventListener('mouseleave', () => {
+        if (!this.ISLOCKED) {
+            popUp.classList.add('hidden');
+            popUp.classList.remove('showing');
+        }
+    });
+}
+
+function changeBoxesState(boxes: NodeListOf<HTMLInputElement>, state: boolean) {
+    boxes.forEach(box => {
+        box.checked = state;
+    });
+}
+
+function getRandomID() {
+    let used: number[] = [];
+
+    function getNumber() {
+        let randomNo = Math.floor(Math.random() * 1000);
+        while (used.includes(randomNo)) {
+            randomNo = Math.floor(Math.random() * 1000);
+        }
+
+        used.push(randomNo);
+        return randomNo;
+    }
+    return getNumber;
+}
+
+function removeChildren(el: HTMLElement) {
+    let child = el.lastElementChild;
+    while (child) {
+        el.removeChild(child);
+        child = el.lastElementChild;
+    }
+}
+
+function clickToAdd() {
```

Comment:
maybe instead of using two functions (clickToAdd and clickToEdit), you can write a toggle function that receives the type of operation (add or remove). This way, you don't duplicate code here.